File: 6. Product of Array Except Self/ProductofArrayExceptSelf.java
import java.util.Arrays;

public class ProductofArrayExceptSelf {
    public static void main(String[] args) {
        int[] nums={1,2,3,4,5};
        System.out.println(Arrays.toString(productExceptSelf(nums)));
    }
        public static int[] productExceptSelf(int[] nums) {
            int zero=0,prod=1,l=nums.length;
            int[] ans=new int[l];
            for(int i=0;i<l;i++)
            {
                if(nums[i]!=0 )
                    prod*=nums[i];
                else{
                    zero++;
                    if(zero>1)
                        return ans;
                }
            }
            for(int i=0;i<l;i++)
            {
                if(zero==0)
                    ans[i]=prod/nums[i];
                if(zero==1 && nums[i]==0)
                    ans[i]=prod;
                // else
                //     nums[i]=0;
            }
            return ans;
        }
    }

File: 6. Product of Array Except Self/ProductOfArrayExceptSelf.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        int front[n];
        int back[n];
        
        front[0] = 1;
        back[n-1] = 1;
        
        for(int i = 1; i < n; i++) {
            front[i] = nums[i-1]*front[i-1];
        }
        
        for(int i = n - 2; i >= 0; i--) {
            back[i] = nums[i+1]*back[i+1];
        }
        
        for(int i = 0; i < nums.size(); i++) {
            ans.push_back(front[i]*back[i]);
        }
        
        return ans;
        
        
    }
};
